src/infer_request.cc

```diff
@@ -154,7 +154,19 @@ InferenceRequest::SetState(InferenceRequest::State new_state)
       break;
     }
     case InferenceRequest::State::EXECUTING: {
-      if (new_state != InferenceRequest::State::RELEASED) {
+      if (new_state == InferenceRequest::State::RELEASED) {
+        // no-op
+      } else if (new_state == InferenceRequest::State::INITIALIZED) {
```

Comment:
I think in one of your side branches I saw the `state_` changed to an atomic variable. I think this issue with callback vs. re-use timing exposed that there can be a race condition on state updates from release callback. Do you think we should keep it atomic for now as long as we "allow" release/re-use to happen in any order?

```
# Expected flow
EXECUTING -> (RELEASED) -> INITIALIZED -> PENDING -> EXECUTING

# Possible flows based on previous (release) callback timing, these can also error out, right?

## ERROR: invalid released->pending
EXECUTING -> INITIALIZED -> (RELEASED) -> PENDING -> EXECUTING

## ERROR: invalid released->executing
EXECUTING -> INITIALIZED -> PENDING -> (RELEASED) -> EXECUTING  
```

Reply:
So we are not enforcing request can only be reused after it is released (or ownership is back to the user)? As we only log warning message and continue the state transition